### ISSUE TITLE
[BUG] Can't infer `ImmutableOrigin` from `MutableOrigin` in function signatures

### ISSUE THREAD

**Issue by martinvuyk** [modular/max#3845]
### Bug description

Implicit constructors for a view type from a mutable origin to an immutable one work, but the origin does not get inferred in function signatures

CC: @lattner, @ConnorGray 

### Steps to reproduce

```mojo
@value
struct SomeType[
    mut: Bool, //,
    T: CollectionElement,
    origin: Origin[mut]
]():
    alias MutSelf = SomeType[T, MutableOrigin.cast_from[origin].result]
    """The mutable version of the SomeType."""
    alias ImmutSelf = SomeType[T, ImmutableOrigin.cast_from[origin].result]
    """The immutable version of the SomeType."""


    @doc_private
    @implicit
    @always_inline("nodebug")
    fn __init__(out self: Self.ImmutSelf, other: Self.MutSelf):
        """Implicitly cast the mutable origin of self to an immutable one.

        Args:
            other: The SomeType to cast.
        """
        self = rebind[Self.ImmutSelf](other)

fn _some_func_1[T: CollectionElement, O: ImmutableOrigin, //](arg_0: SomeType[T, O]):
    print("_some_func_1")

fn _some_func_2[T: CollectionElement, O: ImmutableOrigin, //](arg_0: SomeType[T, O], arg_1: SomeType[T, O]):
    print("_some_func_2")

fn main():
    var a = SomeType[Int, ImmutableAnyOrigin]()
    _some_func_1(a) # works
    _some_func_2(a, a) # works
    
    var b = SomeType[Int, MutableAnyOrigin]()
    _some_func_1[T=Int, O=__type_of(b).ImmutSelf.origin](b) # works
    _some_func_2[T=Int, O=__type_of(b).ImmutSelf.origin](b, b) # works

    _some_func_1(b) # compilation error
    _some_func_2(b, b) # compilation error
```

### System information

```shell
- Provide version information for Mojo by pasting the output of `mojo -v`
`mojo 25.3.0.dev2025040405`
```

**Comment by martinvuyk** [comment]
I think that this would also remove the need for the `@__unsafe_disable_nested_origin_exclusivity` decorator in many functions since they would just be annotated with `ImmutableOrigin` for the readonly ops, and the implicit constructors would then handle any value with a `MutableOrigin` that is passed in

**Comment by martinvuyk** [comment]
ping @lattner. I've updated the description with a smaller repro. I think this issue is important for ergonomics, because IMO casting to immutable should always be inferred.

I am adding the code to implicitly cast to immut to `StringSlice` and `Span` in #3823, I could also add followups for `Pointer` and `UnsafePointer`.

**Comment by lattner** [comment]
Thanks for the head's up, I'll TAL at this tomorrow or early next week.

**Comment by lattner** [comment]
This is super interesting, I would really like this to work, I'll keep poking at it later today

**Comment by lattner** [comment]
It took me awhile to get back to this.  I'm not sure how to make a good error message for this in the short term, but I can explain what is going on. There are two problems:

1) Your init: ` fn __init__(out self: Self.ImmutSelf, other: Self.MutSelf):` can't match the mutable argument, because the type it is looking to match is computed: it can't match "AnyMutableOrigin" against "MutableOrigin.cast_from[origin].result]" because it can't reverse the function.  You don't actually care what the input origin is, so you can use `other: SomeType[T, _]`.

2) If you do that, we can't infer the origin from the source anymore.  The solution to this is to make it so the result type (the out) is determined from the argument type.  

Put together, the solution is to declare it like this:

```
    fn __init__(other: SomeType[T, _], out self: SomeType[T, ImmutableOrigin.cast_from[other.origin].result]):
        """Implicitly cast the mutable origin of self to an immutable one.

        Args:
            other: The SomeType to cast.
        """
        self = rebind[__type_of(self)](other)
```

or:

`fn __init__(other: SomeType[T, _]) -> SomeType[T, ImmutableOrigin.cast_from[other.origin].result]:`

Can you give that a try?

**Comment by lattner** [comment]
@ConnorGray @JoeLoser This pattern is interesting, I think it could be used to get rid of `get_immut()`, by allowing implicit conversions from things like Pointer/Slice etc to their immut version!

**Comment by JoeLoser** [comment]
> [@ConnorGray](https://github.com/ConnorGray) [@JoeLoser](https://github.com/JoeLoser) This pattern is interesting, I think it could be used to get rid of `get_immut()`, by allowing implicit conversions from things like Pointer/Slice etc to their immut version!

FYI @lsh as well since we were recently talking about this. We'll get together as a team and explore this. Thanks for digging into this and minimizing it @martinvuyk!

**Comment by martinvuyk** [comment]
Yessss it wooorked 🎉

**Comment by soraros** [comment]
@lattner So it's the "our unification doesn't wrong backward" kinda problem?

**Comment by lattner** [comment]
I don't think Parameter inference will ever be able to invert a function in general.  Let me give you a simpler example:

```
fn foo[x: Int](arg: T[x+1])

   var abc : T[5]
   foo(abc)
```

Parameter inference would have to be able to figure out how to invert "+1" to "-1" understand that x would have to be inferred to 5-1=4.  There is no reasonable way to implement this form of symbolic reasoning in general.

**Comment by lattner** [comment]
It would be really really really amazing to capture fundamental limitations like ^^ somewhere in a doc so it isn't just stuck in my head. :-)

**Comment by soraros** [comment]
> Parameter inference would have to be able to figure out how to invert "+1" to "-1" understand that x would have to be inferred to 5-1=4. There is no reasonable way to implement this form of symbolic reasoning in general.

Yeah, exactly. I've tried again and again to convince people on Discord that unifying `V[n + m]` with `V[m + n]` is anything but trivial. ~~Unless we embed a SMT solver into the Mojo compiler.~~

> It would be really really really amazing to capture fundamental limitations like ^^ somewhere in a doc so it isn't just stuck in my head. :-)

Please do that. I think we've come to accept quite a few limitations, mostly because our brains are hopelessly marinated in compiler nerdiness—but to most people, they probably sound completely absurd.

**Comment by arthurevans** [comment]
> It would be really really really amazing to capture fundamental limitations like ^^ somewhere in a doc so it isn't just stuck in my head. :-)

What would we call this? Is the general topic limitations of parameter inference?

**Comment by lattner** [comment]
It's more of a "How Parameter Inference Works" topic.
